### DISCUSSION TITLE
Assert function arguments by type instead of name

### DISCUSSION THREAD

**Discussion by haggen** [encode/starlette#1815] [Ideas] [answered]
https://github.com/encode/starlette/blob/f6ea760a80d8b109fb6afd1c03e9a33754e6bb5f/starlette/authentication.py#L29-L38

I'm not a Pythonista so forgive me if it is commonplace in Python, but in general it seems to me be a bad idea to assert that a given function has an argument with a given name. It limits freedom of the developer and makes everything more brittle.

Searching throughout the repository it looks like it's the only instance where it happens.

In my particular use case I was trying to use `req` instead of `request` given the frequency I have to type and reference it.

Would it be acceptable if we validated that the type of the function's arguments instead of their name? 

Let me know and I'll begin working on a PR.

Thanks in advance.

**Comment by Kludex** [comment]
No, sorry.

The conditional would not be removed even if we check the type at runtime, as we avoid breaking changes.

I also think it's out of scope for Starlette.

**Comment by haggen** [comment]
Hey @Kludex thanks for the quick reply.

I don't mean to nag anyone, but if you have the time would you mind answering a few follow-up questions?

Why do you think it would be breaking change? I might be missing something but they way I see it my proposal doesn't break existing code, it just open up to work with more code, namely, functions that match the feature requirements but not necessarily the function's signature.

Why do you think it would be out of scope? I'm not suggesting new features. Not even changing existing features, but only adjusting the implementation of one that right now is limiting what code it is compatible with. Specially if you understand it to be bad practice the way I do.

In any case, I appreciate your decision and you don't have to reply if you don't want to. :) 

Thanks and I hope you have a good day! 🍻

**Reply by Kludex** [reply]
I'm willing to accept a PR about this.

Let's maintain the current behavior, and add the introspection of parameters in case those parameter names are not there.

I'll be closing the issue that I've created from here (#1831), because I don't intend to work on it. PR is welcome tho.

**Comment by Kludex** [comment]
My interpretation of what you said was wrong.

I think it makes sense to think about this. Let's discuss it first.

**Comment by Kludex** [comment, marked as answer]
Created #1831 from this discussion.
